crates/egui/src/containers/tooltip.rs

```diff
@@ -0,0 +1,372 @@
+    pub fn new(widget_id: Id, anchor: impl Into<PopupAnchor>, layer_id: LayerId) -> Self {
+        Self {
+            // TODO(lucasmerlin): Set width somehow (we're missing context here)
+            popup: Popup::new(widget_id, anchor.into(), layer_id)
+                .kind(PopupKind::Tooltip)
+                .gap(4.0)
+                .sense(Sense::hover()),
+            layer_id,
+            widget_id,
+        }
+    }
+
+    /// Show a tooltip for a widget. Always open.
+    pub fn for_widget(response: &Response) -> Self {
+        let popup = Popup::from_response(response)
+            .kind(PopupKind::Tooltip)
+            .gap(4.0)
+            .width(response.ctx.style().spacing.tooltip_width)
+            .sense(Sense::hover());
+        Self {
+            popup,
+            layer_id: response.layer_id,
+            widget_id: response.id,
+        }
+    }
+
+    /// Show a tooltip when hovering an enabled widget.
+    pub fn for_enabled(response: &Response) -> Self {
+        let mut tooltip = Self::for_widget(response);
+        tooltip.popup = tooltip
+            .popup
+            .open(response.enabled() && Self::should_show_tooltip(response));
+        tooltip
+    }
+
+    /// Show a tooltip when hovering a disabled widget.
+    pub fn for_disabled(response: &Response) -> Self {
+        let mut tooltip = Self::for_widget(response);
+        tooltip.popup = tooltip
+            .popup
+            .open(!response.enabled() && Self::should_show_tooltip(response));
+        tooltip
+    }
+
+    /// Show the tooltip at the pointer position.
+    #[inline]
+    pub fn at_pointer(mut self) -> Self {
+        self.popup = self.popup.at_pointer();
+        self
+    }
+
+    /// Set the gap between the tooltip and the anchor
+    ///
+    /// Default: 5.0
+    #[inline]
+    pub fn gap(mut self, gap: f32) -> Self {
+        self.popup = self.popup.gap(gap);
+        self
+    }
+
+    /// Set the layout of the tooltip
+    #[inline]
+    pub fn layout(mut self, layout: Layout) -> Self {
+        self.popup = self.popup.layout(layout);
+        self
+    }
+
+    /// Set the width of the tooltip
+    #[inline]
+    pub fn width(mut self, width: f32) -> Self {
+        self.popup = self.popup.width(width);
+        self
+    }
+
+    /// Show the tooltip
+    pub fn show<R>(
+        self,
+        ctx: &Context,
+        content: impl FnOnce(&mut crate::Ui) -> R,
+    ) -> Option<InnerResponse<R>> {
+        let Self {
+            mut popup,
+            layer_id: parent_layer,
+            widget_id,
+        } = self;
+
+        if !popup.is_open(ctx) {
+            return None;
+        }
+
+        let rect = popup.anchor.rect(popup.id, ctx)?;
+
+        let mut state = ctx.pass_state_mut(|fs| {
+            // Remember that this is the widget showing the tooltip:
+            fs.layers
+                .entry(parent_layer)
+                .or_default()
+                .widget_with_tooltip = Some(widget_id);
```

Comment:
We could maybe get rid of `LayerState::open_popups` if we also stored which widgets has a _popup_ open (attached to them). That would allow us to simplify and improve the "don't show a tooltip if this widget has a popup open" check in https://github.com/emilk/egui/blob/b36311a98ae9475f83aa6d8f5a3a91fe4df5d2b2/crates/egui/src/response.rs#L615-L619

(can be saved for a future PR though)